src/saucelabs/runner.js

```diff
@@ -53,31 +82,58 @@ export default class SauceLabsRunner {
 
         console.log(`See the progress: ${accountUrl} (session: "${this.options.testName}", build: "${this.options.build}")`);
 
-        this.reportingInterval = setInterval(()=>this._report(), REPORTING_TIMEOUT);
+        this.reportingInterval = setInterval(() => this._report(), REPORTING_TIMEOUT);
+    }
+
+    _startJobs (count) {
+        this._getNonRunningJobs()
+            .splice(0, count)
+            .forEach(job => job
+                .run()
+                .then(result => {
+                    this.jobResults = this.jobResults.concat(result);
+
+                    this._startJobs(1);
+                })
+            );
     }
 
-    runTests () {
-        var runner = this;
+    async _runJobs () {
+        var jobsComplete = false;
+        var total        = this.jobs.length;
 
-        var runJobPromises = this.options.browsers.map(function (browser, index) {
-            var job = new Job(runner.options, browser);
+        while (!jobsComplete) {
+            jobsComplete = this._getCompletedJobsCount() + this._getFailedJobsCount() === total;
 
-            runner.jobs.push(job);
-            return wait(index * 1000)
-                .then(function () {
-                    return job.run();
-                });
-        });
+            if (!jobsComplete) {
+                if (this._getNonRunningJobs().length) {
+                    var freeMachinesCount = await this._getFreeMachineCount();
+
+                    if (freeMachinesCount)
+                        this._startJobs(freeMachinesCount);
+                }
+
+                await wait(MACHINE_WAITING_TIMEOUT);
+            }
+        }
+    }
+
```

Comment:
It's all a little complicated. Also I can't get what the difference between `_startJobs` and `_runJobs` methods by their names. It sounds like the same thing. I propose the following structure:
```js
export default class extends EventEmitter {
    constructor () {
        //...
    }

    _onJobDone (job, res) {
        //add result to this.jobResults
        this.emit('job-done');
    }

    _runNextJobs (count) {
        var jobs = getQueuedJobs(count); // get it from this.jobs by state

        if (jobs.length) {
            jobs.forEach(job => {
                job.run()
                    .then(res => this._onJobDone(job, res));
            });
        }
        else
            this._stopFreeMachinesChecker();
    }

    async _checkForFreeMachines () {
        var freeMachineCount = await this._getFreeMachineCount();

        if (freeMachineCount)
            this.emit('free-machines-explored', { count: freeMachineCount });
    }

    _startFreeMachinesChecker () {
        this._checkForFreeMachines();
        this.checkForFreeMachinesInterval = window.setInterval(()=> this._checkForFreeMachines(), INTERVAL);
    }

    _stopFreeMachinesChecker () {
        // ...
    }

    runTests () {
        return new Promise(resolve => {
            //...
            this._runJobs();

            this.on('free-machines-explored', (e) => this._runNextJobs(e.count));
            this.on('job-done', () => this._runNextJobs(1));
            this.on('done', () => resolve(this.jobResults));
        });
    }
}

```